src/realm/object-store/c_api/realm.cpp

```diff
@@ -337,4 +342,32 @@ void CBindingContext::did_change(std::vector<ObserverState> const&, std::vector<
     m_realm_changed_callbacks.invoke();
 }
 
+RLM_API
+realm_thread_observer_token_t*
+realm_set_binding_callback_thread_observer(realm_on_object_store_thread_callback_t on_thread_create,
+                                           realm_on_object_store_thread_callback_t on_thread_destroy,
+                                           realm_on_object_store_error_callback_t on_error, realm_userdata_t userdata,
+                                           realm_free_userdata_func_t free_userdata)
+{
+    realm::c_api::CBindingThreadObserver::ThreadCallback thread_create =
+        [on_thread_create, userdata = UserdataPtr{userdata, free_userdata}]() {
+            on_thread_create(userdata.get());
+        };
+
+    realm::c_api::CBindingThreadObserver::ThreadCallback thread_destroyed =
+        [on_thread_destroy, userdata = UserdataPtr{userdata, free_userdata}]() {
+            on_thread_destroy(userdata.get());
+        };
+
+    realm::c_api::CBindingThreadObserver::ErrorCallback error =
+        [on_error, userdata = UserdataPtr{userdata, free_userdata}](const char* error) {
+            on_error(userdata.get(), error);
+        };
+
+    auto& instance = realm::c_api::CBindingThreadObserver::create();
```

Comment:
`CBindingThreadObserver`  is a singleton, keeping it in memory won't be too expensive, however I am ok if we want to assign a `ptr` to to the token and release it once the token is destroyed.